### PR TITLE
💥 Build using N-API instead of NAN

### LINKED ISSUE #59: Fails to install on Mac OS X
```
$ npm install speaker 
\
> speaker@0.2.5 install /Users/me/Development/app/node_modules/speaker
> node-gyp rebuild

  CC(target) Release/obj.target/output/deps/mpg123/src/output/coreaudio.o
  LIBTOOL-STATIC Release/liboutput.a
  CXX(target) Release/obj.target/binding/src/binding.o
In file included from ../src/binding.cc:8:
In file included from ../src/node_pointer.h:7:
In file included from ../node_modules/nan/nan.h:80:
In file included from ../node_modules/nan/nan_new.h:190:
../node_modules/nan/nan_implementation_12_inl.h:181:66: error: too many arguments to function call, expected at most 2, have 4
  return v8::Signature::New(v8::Isolate::GetCurrent(), receiver, argc, argv);
         ~~~~~~~~~~~~~~~~~~                                      ^~~~~~~~~~
/Users/moogs/.node-gyp/2.4.0/deps/v8/include/v8.h:4188:3: note: 'New' declared here
  static Local<Signature> New(
  ^
1 error generated.
make: *** [Release/obj.target/binding/src/binding.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:269:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 14.3.0
gyp ERR! command "/usr/local/bin/iojs" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /app/node_modules/speaker
gyp ERR! node -v v2.4.0
gyp ERR! node-gyp -v v2.0.1
gyp ERR! not ok 
npm ERR! Darwin 14.3.0
npm ERR! argv "/usr/local/bin/iojs" "/usr/local/bin/npm" "install" "speaker"
npm ERR! node v2.4.0
npm ERR! npm  v2.13.0
npm ERR! code ELIFECYCLE

npm ERR! speaker@0.2.5 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the speaker@0.2.5 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the speaker package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls speaker
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/me/Development/app/npm-debug.log

$ sw_vers
ProductName:    Mac OS X
ProductVersion: 10.10.3
BuildVersion:   14D136

$ node -v
v2.4.0

$ npm -v
2.13.0
```

### LINKED ISSUE #78: fails to install in ubuntu 14.04

### LINKED ISSUE #81: Doesn't compile on Visual Studio 2015 Update 3 : '__pfnDliNotifyHook2': redefinition; different type modifiers
Copy paste from MS's `delayimp.h` header 

``` c++
// Prior to Visual Studio 2015 Update 3, these hooks were non-const.  They were
// made const to improve security (global, writable function pointers are bad).
// If for backwards compatibility you require the hooks to be writable, define
// the macro DELAYIMP_INSECURE_WRITABLE_HOOKS prior to including this header and
// provide your own non-const definition of the hooks.
```

This result in the following error : 

```
win_delay_load_hook.c(35): error C2373: '__pfnDliNotifyHook2': redefinition; different type modifiers
```

Solution, just add the define **DELAYIMP_INSECURE_WRITABLE_HOOKS** in `win_delay_load_hook.c`

``` c++
#define DELAYIMP_INSECURE_WRITABLE_HOOKS
#define WIN32_LEAN_AND_MEAN
#include <windows.h>

#include <delayimp.h>
#include <string.h>
```

### LINKED ISSUE #83: failed install on the macosx
child_process: customFds option is deprecated, use stdio instead.
  CC(target) Release/obj.target/output/deps/mpg123/src/output/coreaudio.o
  LIBTOOL-STATIC Release/liboutput.a
libtool: unrecognized option `-static'
libtool: Try`libtool --help' for more information.
make: **\* [Release/liboutput.a] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/guoxinchun1/.nvm/versions/node/v0.12.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1067:12)
gyp ERR! System Darwin 15.4.0
gyp ERR! command "node" "/Users/guoxinchun1/.nvm/versions/node/v0.12.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/guoxinchun1/project/nodejs/electron/electron-quick-speeker/node_modules/speaker
gyp ERR! node -v v0.12.0
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok 
npm ERR! Darwin 15.4.0
npm ERR! argv "/Users/guoxinchun1/.nvm/versions/node/v0.12.0/bin/node" "/Users/guoxinchun1/.nvm/versions/node/v0.12.0/bin/npm" "install"
npm ERR! node v0.12.0
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE

npm ERR! speaker@0.3.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the speaker@0.3.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the speaker package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls speaker
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/guoxinchun1/project/nodejs/electron/electron-quick-speeker/npm-debug.log

### LINKED ISSUE #106: node-gyp rebuild failed!
trying to update speaker, but it fails..

328 verbose stack Error: speaker@0.3.1 install: `node-gyp rebuild`
328 verbose stack Exit status 1
328 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:280:16)
328 verbose stack     at emitTwo (events.js:125:13)
328 verbose stack     at EventEmitter.emit (events.js:213:7)
328 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
328 verbose stack     at emitTwo (events.js:125:13)
328 verbose stack     at ChildProcess.emit (events.js:213:7)
328 verbose stack     at maybeClose (internal/child_process.js:927:16)
328 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
329 verbose pkgid speaker@0.3.1
330 verbose cwd /Users/xxx/xxx
331 verbose Darwin 16.7.0
332 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "speaker" "--save"
333 verbose node v8.7.0
334 verbose npm  v5.4.2
335 error code ELIFECYCLE
336 error errno 1
337 error speaker@0.3.1 install: `node-gyp rebuild`
337 error Exit status 1
338 error Failed at the speaker@0.3.1 install script.
338 error This is probably not a problem with npm. There is likely additional logging output above.
339 verbose exit [ 1, true ]

### LINKED ISSUE #114: Cannot install Speaker node-gyp
Hello,

I tried to install speaker, but I cannot because of the following issues:

`C:\>node -v
v10.1.0

C:\>npm -v
6.1.0

C:\poc>npm install speaker --save

> speaker@0.4.1 install C:\poc\node_modules\speaker
> node-gyp rebuild

C:\poc\node_modules\speaker>if not defined npm_config_node_gyp (node "C:\Users\<username>\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "node C:\Users\<username>\AppData\Roaming\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
internal/modules/cjs/loader.js:573
    throw err;
    ^

Error: Cannot find module 'C:\poc\node_modules\speaker\node C:\Users\<username>\AppData\Roaming\npm\node_modules\node-gyp\bin\node-gyp.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:571:15)
    at Function.Module._load (internal/modules/cjs/loader.js:497:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:719:10)
    at startup (internal/bootstrap/node.js:228:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:576:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! speaker@0.4.1 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the speaker@0.4.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.`

### LINKED ISSUE #131: crash on windows 10, node 10
At the event  "close", my application exit.
to fix that, i remove "delete ao->device;" (see below) but I'm not sure why .., maybe will leak but working.

NAN_METHOD(Close) {
  Nan::EscapableHandleScope scope;
  audio_output_t *ao = UnwrapPointer<audio_output_t *>(info[0]);
  ao->close(ao);
  int r = 0;
  if (ao->deinit) {
    r = ao->deinit(ao);
  }
  //delete ao->device;
  info.GetReturnValue().Set(scope.Escape(Nan::New<v8::Integer>(r)));
}

### LINKED ISSUE #134: unable to npm install the package
OS: os x 10.14.6 (18G87)

error log:
```
npm install speaker

> speaker@0.4.2 install /Users/x/js/radiorecord/node_modules/speaker
> node-gyp rebuild

  CC(target) Release/obj.target/output/deps/mpg123/src/output/coreaudio.o
  LIBTOOL-STATIC Release/liboutput.a
libtool: unrecognized option `-static'
libtool: Try `libtool --help' for more information.
make: *** [Release/liboutput.a] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:198:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Darwin 18.7.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/x/js/radiorecord/node_modules/speaker
gyp ERR! node -v v10.16.3
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
npm WARN radiorecord@1.0.0 No description
npm WARN radiorecord@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! speaker@0.4.2 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the speaker@0.4.2 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/x/.npm/_logs/2019-09-08T14_03_27_203Z-debug.log
```

### PR DESCRIPTION
**Pull request by LinusU**
closes #133, fixes #134, fixes #131, fixes #114, fixes #106, fixes #83, fixes #81, fixes #78, fixes #59

### CODE REVIEW

**Comment by TooTallNate** [comment]
Wow this is awesome 😄

File: .npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+package-lock=false
```
Comment (TooTallNate) [inline .npmrc]: Meh, we could delete this IMO (and just add `yarn.lock` and `package-lock.json` to the `.gitignore` file).

File: .travis.yml [hunk 2]
```diff
@@ -13,19 +13,7 @@ addons:
...
-script:
-  # Output useful info for debugging
-  - node --version
-  - npm --version
-  # Run tests
-  - npm test
+  - '12'
+  - '10'
+  - '8'
+  - '8.6.0'
```
Comment (TooTallNate) [inline .travis.yml]: Doesn't need to be in this PR, but I would love to switch over to GH Actions instead of AppVeyor/Travis.

File: index.js [hunk 3]
```diff
@@ -7,8 +7,7 @@
 const os = require('os')
 const debug = require('debug')('speaker')
 const binding = require('bindings')('binding')
-const bufferAlloc = require('buffer-alloc')
-const Writable = require('readable-stream/writable')
+const { Writable } = require('readable-stream')
```
Comment (TooTallNate) [inline index.js]: Considering we're dropping support for Node < 8, I think the `buffer-alloc` and `readable-streams` modules can be dumped in favor of their Node core built-in equivalents.

File: package.json [hunk 4]
```diff
@@ -12,18 +12,17 @@
...
-    "mocha": "^3.5.0",
-    "standard": "^10.0.3"
+    "@types/node": "^12.12.14",
+    "buffer-alloc": "^1.1.0",
+    "mocha": "^6.2.2",
+    "standard": "^14.3.1"
   },
   "engines": {
-    "node": ">4"
+    "node": ">=8.6"
```
Comment (TooTallNate) [inline package.json]: Out of curiosity, why `8.6` instead of just `8`?

File: .travis.yml (hunk 2 above)
Comment (TooTallNate) [reply .travis.yml]: I added GH Actions on the master branch. Will need some tweaking for more Node versions once this is merged (for i.e. Node 12).

File: package.json (hunk 4 above)
Comment (LinusU) [reply package.json]: Some things in the N-API, to be honest, I don't exactly remember which function it was 🙈

Comment (LinusU) [reply package.json]: It's:

- `napi_deferred`
- `napi_throw_error` with a third parameter
- `napi_create_string_*` using `NAPI_AUTO_LENGTH`

File: .travis.yml (hunk 2 above)
Comment (LinusU) [reply .travis.yml]: Neat, will complete in a follow up PR 👍

**Comment by LinusU** [comment]
Thanks for the review! I didn't know you were actively looking at this ☺️ 

Since I thought that I was probably going to be the only one to commit I started tagging my commits with [emoji-commit](https://github.com/LinusU/emoji-commit) since I planed on using [emoji-github-releases](https://github.com/LinusU/emoji-github-releases) to automatically generate the changelogs (some examples: [ts-readme-generator/releases](https://github.com/LinusU/ts-readme-generator/releases), [vscode-auto-dark-mode/releases](https://github.com/LinusU/vscode-auto-dark-mode/releases))

We can skip it if you want though!

Otherwise, I'll merge this when the CI is green and then release 0.5.0 :shipit:

**Comment by LinusU** [comment]
Rebased away some redundant commits, merging as soon as CI is green 🚀

**Comment by LinusU** [comment]
[🚢 0.5.0 / 2019-12-07](https://github.com/TooTallNate/node-speaker/releases/tag/v0.5.0)

**Comment by JonSilver** [comment]
Still experiencing #131

**Pull request merged**